lib/Sema/CSGen.cpp

```diff
@@ -1695,9 +1695,18 @@ namespace {
         return Type();
       }
 
-      auto archetype = type->castTo<ArchetypeType>();
-      E->setParamType(archetype);
-      return archetype->getValueType();
+      if (auto archetype = type->getAs<ArchetypeType>()) {
```

Comment:
What else can `type` be here?

Reply:
As far as I know and have tested, it's just `ArchetypeType` and `IntegerType`:

```swift
struct A<let n: Int> {
  func hello() -> Int {
    n // archetype here
  }
}

_ = A<123>.n // integer type here
```